### PR TITLE
Distributed linear regression

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

**Review by Vika-F — COMMENTED** [review]
I would expect that oneCCL and MPI samples are also added when we add a new distributed algorithm.

**Comment by KulikovNikita** [comment]
> I would expect that oneCCL and MPI samples are also added when we add a new distributed algorithm.

@Vika-F , it is fixed!

File: cpp/oneapi/dal/algo/linear_regression/test/spmd.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#include "oneapi/dal/algo/linear_regression/test/spmd_backend_template.hpp"
+
+namespace oneapi::dal::linear_regression::test {
+
+TEMPLATE_LIST_TEST_M(lr_spmd_test, "LR common flow", "[lr][batch]", lr_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->generate(777);
+    this->set_rank_count(GENERATE(2, 4));
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/linear_regression/test/spmd.cpp:26]: I'd also add some odd rank (say 3) for the purpose of testing.

**Review by Vika-F — DISMISSED** [review]
The changes look good to me.
But. I am very sorry, I have overlooked that the changes are in the 'preview' namespace when I had requested to add the samples.
I think we do not add examples and samples for the 'preview' functionality. So the samples I had requested have to be removed. I am sorry again for this.

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp [hunk 2]
```diff
@@ -111,6 +111,21 @@ static train_result<Task> call_dal_kernel(const context_gpu& ctx,
 
     const be::event_vector solve_deps{ last_xty_event, last_xtx_event };
 
+    auto& comm = ctx.get_communicator();
+    if (comm.get_rank_count() > 1) {
+        sycl::event::wait_and_throw(solve_deps);
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:116]: As far as I remember - SPMD tasks can not be executed in truly asynchronous manner. [Proof.](https://github.com/oneapi-src/oneDAL/blob/5c512dd785e2d02270ee49334568ae12b5a2bdbb/cpp/oneapi/dal/backend/communicator.cpp#L198) That's why all SPMD tasks will be executed in more or less sequential order and we will not have any performance benefits.

[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/backend/gpu/train_kernel_norm_eq_dpc.cpp:116]: > The changes look good to me. But. I am very sorry, I have overlooked that the changes are in the 'preview' namespace when I had requested to add the samples. I think we do not add examples and samples for the 'preview' functionality. So the samples I had requested have to be removed. I am sorry again for this.

It is not entirely true. There are samples for pretty much all distributed algorithms. I.e. [this](https://github.com/oneapi-src/oneDAL/blob/5c512dd785e2d02270ee49334568ae12b5a2bdbb/samples/oneapi/dpc/ccl/sources/decision_forest_reg_hist_distr_ccl.cpp#L53) or [this](https://github.com/oneapi-src/oneDAL/blob/5c512dd785e2d02270ee49334568ae12b5a2bdbb/samples/oneapi/dpc/mpi/sources/pca_distr_mpi.cpp#L42)

File: cpp/oneapi/dal/algo/linear_regression/test/spmd.cpp (hunk 1 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/linear_regression/test/spmd.cpp]: Fixed

**Comment by KulikovNikita** [comment]
> The changes look good to me. But. I am very sorry, I have overlooked that the changes are in the 'preview' namespace when I had requested to add the samples. I think we do not add examples and samples for the 'preview' functionality. So the samples I had requested have to be removed. I am sorry again for this.

It is not entirely true. There are samples for pretty much all distributed algorithms. I.e. [this](https://github.com/oneapi-src/oneDAL/blob/5c512dd785e2d02270ee49334568ae12b5a2bdbb/samples/oneapi/dpc/ccl/sources/decision_forest_reg_hist_distr_ccl.cpp#L53) or [this](https://github.com/oneapi-src/oneDAL/blob/5c512dd785e2d02270ee49334568ae12b5a2bdbb/samples/oneapi/dpc/mpi/sources/pca_distr_mpi.cpp#L42)

**Comment by KulikovNikita** [comment]
@Vika-F Please review

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/ed7aee41-8da4-f156-8bf5-a4bf010d0e2e

**Comment by ethanglaser** [comment]
It seems like the tests/spmd.cpp is much shorter than other distributed algos (ie [DBSCAN](https://github.com/KulikovNikita/daal/blob/dev/linreg-spmd/cpp/oneapi/dal/algo/dbscan/test/spmd.cpp), [covariance](https://github.com/KulikovNikita/daal/blob/dev/linreg-spmd/cpp/oneapi/dal/algo/covariance/test/spmd.cpp)) - is there a reason for this?

**Comment by KulikovNikita** [comment]
> It seems like the tests/spmd.cpp is much shorter than other distributed algos (ie [DBSCAN](https://github.com/KulikovNikita/daal/blob/dev/linreg-spmd/cpp/oneapi/dal/algo/dbscan/test/spmd.cpp), [covariance](https://github.com/KulikovNikita/daal/blob/dev/linreg-spmd/cpp/oneapi/dal/algo/covariance/test/spmd.cpp)) - is there a reason for this?

1. It is much simpler algorithm
2. Pretty much entire test functionality is shared between batch & spmd. It means that `fixture.hpp` is containing all this parameters

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/ed7f908d-da37-f1f1-8511-a4bf010d0e2e

File: samples/oneapi/dpc/ccl/sources/linear_regression_distr_ccl.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
```
Comment (Alexandr-Solovev) [inline samples/oneapi/dpc/ccl/sources/linear_regression_distr_ccl.cpp:2]: Please, change copyright year

File: samples/oneapi/dpc/mpi/sources/linear_regression_distr_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
```
Comment (Alexandr-Solovev) [inline samples/oneapi/dpc/mpi/sources/linear_regression_distr_mpi.cpp:2]: same comment

**Review by Alexandr-Solovev — APPROVED** [review]

File: samples/oneapi/dpc/ccl/sources/linear_regression_distr_ccl.cpp (hunk 3 above)
Comment (KulikovNikita) [reply samples/oneapi/dpc/ccl/sources/linear_regression_distr_ccl.cpp]: Fixed

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**